sway-lsp/src/capabilities/code_actions/mod.rs

```diff
@@ -0,0 +1,193 @@
+pub mod abi_decl;
+pub mod struct_decl;
+
+use crate::core::{
+    session::Session,
+    token::{Token, TypedAstToken},
+    token_map::TokenMap,
+};
+pub use crate::error::DocumentError;
+use serde_json::Value;
+use std::{collections::HashMap, sync::Arc};
+use sway_core::{language::ty::TyDeclaration, transform::AttributesMap, Engines, TypeParameter};
+use sway_types::Spanned;
+use tower_lsp::lsp_types::{
+    CodeAction, CodeActionDisabled, CodeActionKind, CodeActionOrCommand, CodeActionResponse,
+    Position, Range, TextDocumentIdentifier, TextEdit, Url, WorkspaceEdit,
+};
+
+pub(crate) const CODE_ACTION_IMPL_TITLE: &str = "Generate impl for";
+pub(crate) const CODE_ACTION_NEW_TITLE: &str = "Generate `new`";
+pub(crate) const CONTRACT: &str = "Contract";
+pub(crate) const TAB: &str = "    ";
+
+#[derive(Clone)]
+pub(crate) struct CodeActionContext<'a> {
+    engines: Engines<'a>,
+    tokens: &'a TokenMap,
+    token: &'a Token,
+    uri: &'a Url,
+}
+
+pub(crate) fn code_actions(
+    session: Arc<Session>,
+    range: &Range,
+    text_document: TextDocumentIdentifier,
+    temp_uri: &Url,
+) -> Option<CodeActionResponse> {
+    let (_, token) = session
+        .token_map()
+        .token_at_position(temp_uri, range.start)?;
+    let type_engine = session.type_engine.read();
+    let decl_engine = session.decl_engine.read();
+    let ctx = CodeActionContext {
+        engines: Engines::new(&type_engine, &decl_engine),
+        tokens: session.token_map(),
+        token: &token.clone(),
+        uri: &text_document.uri,
+    };
+    token.typed.and_then(|typed_token| match typed_token {
+        TypedAstToken::TypedDeclaration(decl) => match decl {
+            TyDeclaration::AbiDeclaration(ref decl_id) => abi_decl::code_actions(decl_id, ctx),
+            TyDeclaration::StructDeclaration(ref decl_id) => {
+                struct_decl::code_actions(decl_id, ctx)
+            }
+            _ => None,
+        },
+        _ => None,
+    })
+}
+
+pub(crate) trait CodeActionTrait<'a, T: Spanned> {
```

Comment:
```suggestion
pub(crate) trait CodeAction<'a, T: Spanned> {
```
 the fact it's a trait is implied by the type.